mlflow/pyfunc/__init__.py

```diff
@@ -1046,17 +1054,65 @@ def _predict_row_batch(predict_fn, args):
         )
         result = predict_fn(pdf)
 
+        if isinstance(result, dict):
+            result = {k: list(v) for k, v in result.items()}
+
         if not isinstance(result, pandas.DataFrame):
             result = pandas.DataFrame(data=result)
 
+        spark_primitive_type_to_np_type = {
+            IntegerType: np.int32,
+            LongType: np.int64,
+            FloatType: np.float32,
+            DoubleType: np.float64,
+            BooleanType: bool,
+        }
+
+        if isinstance(result_type, SparkStructType):
+            result_dict = {}
+            for field_name in result_type.fieldNames():
+                field_type = result_type[field_name].dataType
+                field_values = result[field_name]
+
+                if type(field_type) in spark_primitive_type_to_np_type:
+                    np_type = spark_primitive_type_to_np_type[type(field_type)]
+                    field_values = field_values.astype(np_type)
+
+                elif type(field_type) == ArrayType:
+                    elem_type = field_type.elementType
+                    if type(elem_type) not in spark_primitive_type_to_np_type:
+                        raise MlflowException(
+                            "Unsupported array type field with element type "
+                            f"{elem_type.simpleString()} in struct type.",
+                            error_code=INVALID_PARAMETER_VALUE,
+                        )
+                    np_type = spark_primitive_type_to_np_type[type(elem_type)]
+
+                    field_values = [
+                        np.array(v, dtype=np.dtype(type(elem_type))) for v in field_values
+                    ]
+
+                else:
```

Comment:
Are there use cases for a StructField of StructType within a StructType for certain models? Is this recursive inspection useful for the future?

Reply:
I think we don't need this case for now. If user has the use case, we can add it in future.